peak/float.py

```diff
@@ -1,38 +1,424 @@
+from peak.black_box import BlackBox
+from hwtypes import BitVector, Bit
+from hwtypes.adt import Enum
+import magma
+
+class RoundingMode(Enum):
+    RNE = 0
+    RTZ = 1
+    RDN = 2
+    RUP = 3
+    RMM = 4
+
+def RoudningMode_utils(family):
+    RM = RoundingMode
+    RM_c = family.get_constructor(RoundingMode)
+    BV3 = lambda e: family.BitVector[3](e.value)
+    def from_rm(rm: RoundingMode) -> BitVector[3]:
+        return (rm==RM.RNE).ite(
+            BV3(RM.RNE),
+            (rm==RM.RTZ).ite(
+                BV3(RM.RTZ),
+                (rm == RM.RDN).ite(
+                    BV3(RM.RDN),
+                    (rm == RM.RUP).ite(
+                        BV3(RM.RUP),
+                        BV3(RM.RMM),
+                    ),
+                ),
+            ),
+        )
+
+    def to_rm(rm: BitVector[3]) -> RoundingMode:
+        return (rm==BV3(RM.RNE)).ite(
+            RM_c(RM.RNE),
+            (rm==BV3(RM.RTZ)).ite(
+                RM_c(RM.RTZ),
+                (rm == BV3(RM.RDN)).ite(
+                    RM_c(RM.RDN),
+                    (rm == BV3(RM.RUP)).ite(
+                        RM_c(RM.RUP),
+                        RM_c(RM.RMM),
+                    ),
+                ),
+            ),
+        )
+
+    return SimpleNamespace(**locals())
 
 @lru_cache(None)
-def Float(frac, exp):
-    width = frac + exp + 1
+def float_lib_gen(exp_bits: int, frac_bits: int):
+
+    width = 1 + exp_bits + frac_bits
     Data = BitVector[width]
 
+    #Returns a hwtypes float type
     @family_closure
-    def add_fc(family: TypeFamily):
+    def Float_fc(family):
+        if isinstance(family, MagmaFamily):
+            Float = magma.BFloat[width]
+            Float.reinterpret_from_bv = lambda bv: Float(bv)
+            Float.reinterpret_as_bv = lambda f: magma.Bits[width](f)
+            return Float
+        elif isinstance(family, SMTFamily):
+            return SMTFPVector[exp_bits, frac_bits, RM.RNE, False]
+        else:
+            return FPVector[exp_bits, frac_bits, RM.RNE, False]
+
+    @family_closure
+    def Abs_fc(family: TypeFamily):
+        Float = Float_fc(family)
+
         @family.assemble(locals(), globals())
-        class add(Peak, BlackBox):
+        class Abs(Peak, BlackBox):
+            def __call__(self, in_: Data) -> Data:
+                ...
+        return Abs
+
+    @family_closure
+    def Neg_fc(family: TypeFamily):
+        Float = Float_fc(family)
+
+        @family.assemble(locals(), globals())
+        class Neg(Peak, BlackBox):
+            def __call__(self, in_: Data) -> Data:
+                ...
+        return Neg
+
+    @family_closure
+    def Add_fc(family: TypeFamily):
+        Float = Float_fc(family)
+        @family.assemble(locals(), globals())
+        class Add(Peak, BlackBox):
+            def __call__(self, rm: RoundingMode, in0: Data, in1: Data) -> Data:
+                in0_f = Float.reinterpret_from_bv(in0)
```

Comment:
Should define Float in terms of the RM

Reply:
Also could use a decorator to do all the casting:
```
def cast_binary(f):
    def wrapper(self, rm, in0, in1):
        Float = family.FPVector[exp_bits, frac_bits, rm_to_fp_rm(rm), False] # optional
        in0_f = Float.reinterpet_from_bv(in0)
        in1_f = Float.reinterpet_from_bv(in0)
        out = f(self, rm, in0_f, in1_f)
        return Float.reinterpret_as_bv(out)
    return wrapper

...
@family.assemble(...)
class Add(...):
   @cast_binary
   def __call__(self, rm, in0, in1):
       return in0 + in1
```

Reply:
Might be possible (assuming that family.assemble doesnt complain). if the others need to be implemented at some point, Ill do this.

Reply:
Assuming you put the right annotations on `wrapper` and `wrapper` doesn't need to be ssa'd then a decorator should not be problematic.

Reply:
Or actually the decorator might be ssa'd so it will need  be okay with that, and that wrapped function will likely not be ssa'd.  In any event for this use case I think it should be fine.